job-server/src/spark.jobserver/AkkaClusterSupervisorActor.scala

```diff
@@ -0,0 +1,217 @@
+package spark.jobserver
+
+import java.io.IOException
+import java.util.concurrent.TimeUnit
+
+import akka.actor._
+import akka.cluster.Cluster
+import akka.cluster.ClusterEvent.{MemberUp, MemberEvent, InitialStateAsEvents}
+import akka.util.Timeout
+import com.typesafe.config.Config
+import ooyala.common.akka.InstrumentedActor
+import spark.jobserver.util.SparkJobUtils
+import scala.collection.mutable
+import scala.util.{Try, Success, Failure}
+import scala.sys.process._
+
+/**
+ * Created by ankits on 4/7/15.
+ */
+class AkkaClusterSupervisorActor(daoActor: ActorRef) extends InstrumentedActor {
+  import ContextSupervisor._
+  import scala.collection.JavaConverters._
+  import scala.concurrent.duration._
+
+  val config = context.system.settings.config
+  val defaultContextConfig = config.getConfig("spark.context-settings")
+  val contextInitTimeout = config.getDuration("spark.context-settings.context-init-timeout",
+                                                TimeUnit.SECONDS)
+  val managerStartCommand = config.getString("deploy.manager-start-cmd")
+  import context.dispatcher
+
+  //actor name -> (context name, context config, context isadhoc)
+  private val contextInitInfos = mutable.HashMap.empty[String, (String, Config, Boolean)]
+  //actor name -> (success callback, failure callback)
+  private val contextCallbacks = mutable.HashMap.empty[String, (ActorRef => Unit, Throwable => Unit)]
+
+  private val contexts = mutable.HashMap.empty[String, ActorRef]
+  private val resultActors = mutable.HashMap.empty[String, ActorRef]
```

Comment:
Maybe we should have a Map[String, (ActorRef, ActorRef)] (contextActor, resultActor) ? they kind of work together